server/src/main/java/org/elasticsearch/plugins/PluginsLoader.java

```diff
@@ -284,9 +323,13 @@ static LayerAndLoader createPlugin(
                     Set.of("org.elasticsearch.server") // TODO: instead of denying server, allow only jvm + stable API modules
                 )
             );
+            nonModularPluginDescriptorByClassLoader.put(layerAndLoader.loader(), plugin);
+            return layerAndLoader;
         } else {
             logger.debug(() -> "Loading bundle: " + plugin.getName() + ", non-modular");
-            return LayerAndLoader.ofLoader(URLClassLoader.newInstance(bundle.urls.toArray(URL[]::new), pluginParentLoader));
+            var layerAndLoader = LayerAndLoader.ofLoader(URLClassLoader.newInstance(bundle.urls.toArray(URL[]::new), pluginParentLoader));
+            nonModularPluginDescriptorByClassLoader.put(layerAndLoader.loader(), plugin);
```

Comment:
For plugins that have no separate `ModuleLayer` (or aren't modular at all), I think that key-ing them by ClassLoader could be a solution? We know each plugin has a distinct ClassLoader.

Reply:
We'll leave this as a backup plan; for now, this PR will only implement resolution for modular plugins and stable plugins (which have a synthetic Module and ModuleLayer).
There is WIP to extend synthetic Module and ModuleLayer to all plugins; if that succeed, we'll not need anything else. If not, we can come back and add the ClassLoader fallback later.